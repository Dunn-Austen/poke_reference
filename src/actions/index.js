export const cacheNames = pokeNames => ({
  type: 'CACHE_NAMES',
  pokeNames
});

export const cacheTypes = pokeTypes => ({
  type: 'CACHE_TYPES',
  pokeTypes
});

export const handleError = errorMessage => ({
  type: 'HANDLE_ERROR',
  errorMessage
})

export const isLoading = loadingStatus => ({
  type: 'IS_LOADING',
  loadingStatus
})

export const storePokemon = pokeData => ({
  type: 'STORE_POKEMON',
  pokeData
})

export const storeOpponentTypes = opponentTypes => ({
  type: 'STORE_OPPONENTS',
  opponentTypes
})
