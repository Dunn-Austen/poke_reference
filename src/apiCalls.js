export const fetchEmAll = () => {
  return fetch('https://pokeapi.co/api/v2/pokemon/?limit=4000')
    .then(response => {
      if (!response.ok) {
        throw Error('Error with catchEmAll get attempt')
      }
      return response.json()
    })
}

export const fetchTypes = () => {
  return fetch('https://pokeapi.co/api/v2/type')
    .then(response => {
      if (!response.ok) {
        throw Error('Error with fetchTypes')
      }
      return response.json()
    })
}
