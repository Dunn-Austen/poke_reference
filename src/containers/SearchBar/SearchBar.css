.search-container {
  align-items: center;
  background-color: #6f7574;
  border: 2px ridge red;
  box-shadow: 0px 0px 6px 4px black;
  display: flex;
  flex-direction: column;
  height: 210px;
  justify-content: center;
  width: 370px;
}

.search-prompt {
  color: white;
  font-size: 36px;
  margin: 0;
  text-shadow: 3px 4px 1px black;
}

#pokefoe-input {
  color: black;
  font-size: 15px;
  height: 16px;
  text-align: center;
}

label {
  color: white;
  font-size: 18px;
  margin-bottom: 7px;
  text-shadow: 2px 3px 1px black;
}

.search-btn {
  border-radius: 100px;
  font-size: 18px;
  font-weight: bold;
  height: 30px;
  margin: 15px 0;
  width: 80px;
}

.search-btn:hover {
  background-color: #8C7B29;
  color: white;
}

.error {
  color: #ea2323;
  font-weight: bolder;
  text-shadow: 0px 4px 6px black;
}
