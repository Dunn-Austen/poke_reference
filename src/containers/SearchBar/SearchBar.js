import React, { Component } from 'react';
import './SearchBar.css';
import { fetchPokemonData, fetchTypeData, fetchOpponentTypeData } from '../../apiCalls';
import { storePokemon, cacheTypes, handleError, isLoading, storeOpponentTypes } from '../../actions';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';

export class SearchBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      inputValue: '',
      error: '',
    };
  }

  validateInputs = () => {
    const { pokeNames } = this.props;
    const allNames = pokeNames.map(pokemon => {
      return pokemon.name
    });
      if (!allNames.includes(this.state.inputValue)) {
        this.setState({error: 'Pokemon not found / Spelling error'})
      }
  }

  handleClick(inputValue) {
    const { storePokemon, cacheTypes, handleError, isLoading, pokeTypes, storeOpponentTypes } = this.props;
    this.validateInputs();
    fetchPokemonData(inputValue)
      .then(data => {
        storePokemon(data)
        fetchTypeData(data)
          .then(data => {
            cacheTypes(data)
            fetchOpponentTypeData(data)
              .then(data => {
                storeOpponentTypes(data)
              })
              .catch(error => {
                console.log('Error with opponentTypes')
              })
          })
          .catch(error => {
            console.log('Error with pokeTypes retrieval')
          })
        })
      .catch(error => {
        handleError('Server error. Try again shortly')
      })
  }

  handleInputChange(event) {
    const { handleError } = this.props;
    this.setState({error: ''});
    handleError('');
    this.setState({ inputValue: event.target.value.toLowerCase() });
  }

  render () {
    const { pokeData, errorMessage } = this.props;
    return pokeData.name ? (
        <Redirect to='/pokepage' />
      ) : (
      <section className='search-container'>
        <h1 className='search-prompt'>Find your Foe</h1>
        {this.state.error !== '' &&
          <h1 className='error'>
            {this.state.error}
          </h1>
        }
        {this.state.error === '' &&
          <h1 className='error'>
            {errorMessage}
          </h1>
        }
        <label htmlFor='pokefoe-input'>Pokemon Name</label>
        <input
          id='pokefoe-input'
          type='text'
          placeholder='Mind your spelling'
          onChange={(event) => {this.handleInputChange(event)}}
        />
        <button
          className="search-btn"
          onClick={(event) => {this.handleClick(this.state.inputValue)}}
        >
          Search
        </button>
      </section>
    )
  }
}

export const mapStateToProps = state => ({
  pokeData: state.pokeData,
  pokeNames: state.pokeNames,
  errorMessage: state.errorMessage,
  pokeTypes: state.pokeTypes
})

export const mapDispatchToProps = dispatch => ({
  storePokemon: pokeData => dispatch(storePokemon(pokeData)),
  cacheTypes: pokeTypes => dispatch(cacheTypes(pokeTypes)),
  handleError: errorMessage => dispatch(handleError(errorMessage)),
  isLoading: loadingStatus => dispatch(isLoading(loadingStatus)),
  storeOpponentTypes: opponentTypes => dispatch(storeOpponentTypes(opponentTypes))
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchBar);

SearchBar.propTypes = {
  pokeData: PropTypes.object,
  storePokemon: PropTypes.func,
  cacheTypes: PropTypes.func,
  handleError: PropTypes.func,
  isLoading: PropTypes.bool,
  storeOpponentTypes: PropTypes.func,
  pokeNames: PropTypes.array,
  errorMessage: PropTypes.string,
  pokeTypes: PropTypes.array
}
