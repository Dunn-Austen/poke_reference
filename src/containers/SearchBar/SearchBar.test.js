import React from 'react';
import { shallow } from 'enzyme';
import { SearchBar, mapStateToProps, mapDispatchToProps } from './SearchBar';
import { cacheNames, cacheTypes, handleError, isLoading, storePokemon, storeOpponentTypes } from '../../actions';
import { fetchPokemonData, fetchTypeData, fetchOpponentTypeData  } from './../../apiCalls';

jest.mock('../../apiCalls.js');

describe('SearchBar', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<SearchBar
      pokeData={{
        abilities: [],
        height: 10,
        name: 'gengar',
        types: [],
        sprites: {
          back_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/1.png",
          back_shiny: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/shiny/1.png",
          front_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
          front_shiny: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/shiny/1.png"
        }
      }}
      pokeTypes={[{damage_relations: {double_damage_from: []}, pokemon: [{pokemon:{name: "abra", url: "https://pokeapi.co/api/v2/pokemon/63/"}}]}]}
      opponentTypes={[{damage_relations: {double_damage_from: []}, pokemon: [{pokemon:{name: "abra", url: "https://pokeapi.co/api/v2/pokemon/63/"}}]}]}
      storePokemon={jest.fn()}
      cacheTypes={jest.fn()}
      handleError={jest.fn()}
      isLoading={jest.fn()}
      storeOpponentTypes={jest.fn()}
      pokeNames={['abra', 'pikachu']}
      errorMessage={'error'}
    />);

    fetchPokemonData.mockImplementation(() => {
      return Promise.resolve({abilities: [], height: 10, name: 'gengar', types: []})
    });

    fetchTypeData.mockImplementation(() => {
      return Promise.resolve([{damage_relations: {}, id: 15, name: 'ice', pokemon: []}])
    });

    fetchOpponentTypeData.mockImplementation(() => {
      return Promise.resolve([{damage_relations: {}, id: 15, name: 'ice', pokemon: []}])
    });

  });

  it('should match the SearchBar Snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should invoke handleClick on button click', () => {
    const mockArgument = 'pikachu';
    wrapper.instance().handleClick = jest.fn();
    // wrapper.find('button').simulate('click', mockArgument);
    wrapper.instance().handleClick();

    expect(wrapper.instance().handleClick).toHaveBeenCalled();
  });

  // it('should invoke handleClick on button click', () => {
  //   const mockArgument = 'pikachu';
  //   wrapper.instance().handleClick = jest.fn();
  //   wrapper.find('button').simulate('click', mockArgument);
  //
  //   expect(wrapper.instance().handleClick).toHaveBeenCalledWith(mockArgument);
  // });

  it('should invoke validateInputs when handleClick is called', () => {
    const mockArgument = 'pikachu';
    wrapper.instance().validateInputs = jest.fn();
    wrapper.instance().handleClick(mockArgument);

    expect(wrapper.instance().validateInputs).toHaveBeenCalled();
  });

  it('should setState when validateInputs when the spelling fails to match the names list',
  () => {
    const initialState = {
      inputValue: '',
      error: ''
    };
    const alteredState = {
      inputValue: '',
      error: 'Pokemon not found / Spelling error'
    };
    wrapper.setState(initialState);
    expect(wrapper.state()).toEqual(initialState);

    wrapper.instance().validateInputs();
    expect(wrapper.state()).toEqual(alteredState);
});


  describe('mapsStateToProps', () => {
    it('should return only the pertinent information from the redux store', () => {
      const mockState = {
        pokeNames: [{pokemon: 'alakazam', url: 'url'}],
        pokeTypes: [],
        errorMessage: '',
        isLoading: false,
        pokeData: {}
      };
      const expected = {
        pokeData: {},
        pokeNames: [{pokemon: 'alakazam', url: 'url'}],
        errorMessage: '',
        pokeTypes: []
      };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });



  describe('mapDispatchToProps', () => {
    it('calls dispatch with a cacheTypes action when cacheNames is called',
      () => {
        const mockDispatch = jest.fn();
        const actionToDispatch = cacheTypes([{name: 'poison'}]);
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.cacheTypes([{name: 'poison'}]);

        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with a storePokemon action when storePokemon is called',
      () => {
        const mockDispatch = jest.fn();
        const actionToDispatch = storePokemon([{name: 'pikachu'}]);
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.storePokemon([{name: 'pikachu'}]);

        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with a handleError action when handleError is called',
      () => {
        const mockDispatch = jest.fn();
        const actionToDispatch = handleError('error');
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.handleError('error');

        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with an isLoading action when isLoading is called',
      () => {
        const mockDispatch = jest.fn();
        const actionToDispatch = isLoading(false);
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.isLoading(false);

        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with an storeOpponentTypes action when storeOpponentTypes is called',
      () => {
        const mockDispatch = jest.fn();
        const actionToDispatch = storeOpponentTypes([{type: {name: 'poison'}}, {}]);
        const mappedProps = mapDispatchToProps(mockDispatch);

        mappedProps.storeOpponentTypes([{type: {name: 'poison'}}, {}]);

        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
})
