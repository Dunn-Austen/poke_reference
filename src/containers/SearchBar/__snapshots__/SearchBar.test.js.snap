// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match the NavBar Snapshot 1`] = `undefined`;

exports[`SearchBar should match the SearchBar Snapshot 1`] = `
<Redirect
  to="/pokepage"
/>
`;
